# Oily Giant New Well Prediction

## Download and prepare data

In [1]:
import pandas as pd

region_1 = pd.read_csv('geo_data_0.csv')

print("Region 1")
print("--------")
print(region_1.head())
print(region_1.info())

Region 1
--------
      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [2]:
region_2 = pd.read_csv('geo_data_1.csv')

print("Region 2")
print("--------")
print(region_2.head())
print(region_2.info())

Region 2
--------
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [3]:
region_3 = pd.read_csv('geo_data_2.csv')

print("Region 3")
print("--------")
print(region_3.head())
print(region_3.info())

Region 3
--------
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


I downloaded each data from and put them each into a sperate data frame for further calculations. I printed the first five rows in the data frame info to make sure there were no null values and to determine which values will be my features and my target. 

## Predicting oil reserves for each region.

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error


### Region 1

In [5]:
r1_train, r1_valid = train_test_split(region_1, test_size=0.25, random_state=12345)

r1_train_features = r1_train[['f0', 'f1', 'f2']]
r1_train_target = r1_train['product']
r1_valid_features = r1_valid[['f0', 'f1', 'f2']]
r1_valid_target = r1_valid['product']

r1_model = LinearRegression()
r1_model.fit(r1_train_features, r1_train_target)
r1_predicted_valid = r1_model.predict(r1_valid_features)

r1_mse = mean_squared_error(r1_valid_target, r1_predicted_valid)
r1_avg_volume = r1_predicted_valid.mean()

print("Region 1 Avg volume =", r1_avg_volume)
print("Region 1 RMSE =", r1_mse ** 0.5)

Region 1 Avg volume = 92.59256778438038
Region 1 RMSE = 37.5794217150813


### Region 2

In [6]:
r2_train, r2_valid = train_test_split(region_2, test_size=0.25, random_state=12345)

r2_train_features = r2_train[['f0', 'f1', 'f2']]
r2_train_target = r2_train['product']
r2_valid_features = r2_valid[['f0', 'f1', 'f2']]
r2_valid_target = r2_valid['product']

r2_model = LinearRegression()
r2_model.fit(r2_train_features, r2_train_target)
r2_predicted_valid = r2_model.predict(r2_valid_features)

r2_mse = mean_squared_error(r2_valid_target, r2_predicted_valid)
r2_avg_volume = r2_predicted_valid.mean()

print("Region 2 Avg volume =", r2_avg_volume)
print("Region 2 RMSE =", r2_mse ** 0.5)

Region 2 Avg volume = 68.72854689544602
Region 2 RMSE = 0.8930992867756158


### Region 3

In [7]:
r3_train, r3_valid = train_test_split(region_3, test_size=0.25, random_state=12345)

r3_train_features = r3_train[['f0', 'f1', 'f2']]
r3_train_target = r3_train['product']
r3_valid_features = r3_valid[['f0', 'f1', 'f2']]
r3_valid_target = r3_valid['product']

r3_model = LinearRegression()
r3_model.fit(r3_train_features, r3_train_target)
r3_predicted_valid = r3_model.predict(r3_valid_features)

r3_mse = mean_squared_error(r3_valid_target, r3_predicted_valid)
r3_avg_volume = r3_predicted_valid.mean()

print("Region 3 Avg volume =", r3_avg_volume)
print("Region 3 RMSE =", r3_mse ** 0.5)

Region 3 Avg volume = 94.96504596800489
Region 3 RMSE = 40.02970873393434


For this task I split the datafor each region into a training and validation set, then defined my festues and target for each set. I then trained three seperate models, one for each region. I predicted the average volume and determined the RMSE for each region. The third region had the highest average volume, however the RMSE was very high. This means that the region can be very unpredictable in terms of volume of reserves in the well. Region number 2 Preformed the best in terms of RMSE meaning the average volume is very predictable.

## Prepare for profit calculation

In [22]:

r1_revenue = r1_avg_volume * 4500
r2_revenue = r2_avg_volume * 4500
r3_revenue = r3_avg_volume * 4500

revenue_per_unit = 4500
cost = 100000000
cost_per_well = 100000000 / 200

print("Cost per well =", cost_per_well)
print("Minimum volume for profit (per unit) =", ((cost / 200) / revenue_per_unit))


Cost per well = 500000.0
Minimum volume for profit (per unit) = 111.11111111111111


I found that the minimum volume for a well to be considered profitable is about 112,000 barrels. The cost to build one well is about 500,000 USD 112,000 barrels will bring in 504,000 USD in revenue.

## Calculate profit from a set of selected oil wells and model predictions

In [24]:
def well_profit(target_valid, predicted_valid):
    predicted_valid = pd.Series(predicted_valid, index=target_valid.index)
    sample = predicted_valid.sample(n=500, random_state=12345)
    top_200 = sample.nlargest(200)
    target_200 = target_valid.loc[top_200.index]
    profit = target_200.sum() * revenue_per_unit - cost
    return profit

r1_profit = well_profit(r1_valid_target, r1_predicted_valid)
r2_profit = well_profit(r2_valid_target, r2_predicted_valid)
r3_profit = well_profit(r3_valid_target, r3_predicted_valid)

print("Region 1 projected profit =", r1_profit)
print("Region 2 projected profit =", r2_profit)
print("Region 3 projected profit =", r3_profit)

Region 1 projected profit = 6790688.578924239
Region 2 projected profit = 7794798.84176816
Region 3 projected profit = 4399901.430221811


I created a function that will find the projected profit of each region by taking the top 200 wells in each regions predictions and I found the the second region had the highest projected profit. I would reccomed the second region based off of this calculation.

## Calculate risks and profit for each region

### Region 1

In [17]:
import numpy as np

state = np.random.RandomState(12345)
    
r1_values = []
for i in range(1000):
    target_subsample = r1_valid_target.sample(n=500, replace=True, random_state=state)
    pred_subsample = r1_predicted_valid = pd.Series(r1_predicted_valid, index=r1_valid_target.index)
    
    r1_values.append(well_profit(target_subsample, pred_subsample))
    
r1_values = pd.Series(r1_values)
mean = r1_values.mean()
print("Average profit =", mean)

r1_lower = r1_values.quantile(0.025)
r1_upper = r1_values.quantile(0.975)
print("95% confidence interval:", r1_lower, "-", r1_upper)

r1_losses = r1_values[r1_values < 0].count()
r1_loss_prob = r1_losses / 1000
print("Probabillity of a loss: {:.1%}".format(r1_loss_prob))

Average profit = 6007352.442611662
95% confidence interval: 129483.3113511391 - 12311636.057914983
Probabillity of a loss: 2.0%


### Region 2

In [18]:
state = np.random.RandomState(12345)
r2_values = []
for i in range(1000):
    target_subsample = r2_valid_target.sample(n=500, replace=True, random_state=state)
    pred_subsample = r2_predicted_valid = pd.Series(r2_predicted_valid, index=r2_valid_target.index)
    
    r2_values.append(well_profit(target_subsample, pred_subsample))
    
r2_values = pd.Series(r2_values)
mean = r2_values.mean()
print("Average profit =", mean)

r2_lower = r2_values.quantile(0.025)
r2_upper = r2_values.quantile(0.975)
print("95% confidence interval:", r2_lower, ",", r2_upper)

r2_losses = r2_values[r2_values < 0].count()
r2_loss_prob = r2_losses / 1000
print("Probabillity of a loss: {:.1%}".format(r2_loss_prob))

Average profit = 6652410.5822107205
95% confidence interval: 1579884.8131991457 , 11976415.874631526
Probabillity of a loss: 0.3%


### Region 3

In [12]:
state = np.random.RandomState(12345)
r3_values = []
for i in range(1000):
    target_subsample = r3_valid_target.sample(n=500, replace=True, random_state=state)
    pred_subsample = r3_predicted_valid = pd.Series(r3_predicted_valid, index=r3_valid_target.index)
    
    r3_values.append(well_profit(target_subsample, pred_subsample))
    
r3_values = pd.Series(r3_values)
mean = r3_values.mean()
print("Average profit =", mean)

r3_lower = r3_values.quantile(0.025)
r3_upper = r3_values.quantile(0.975)
print("95% confidence interval :", r3_lower, "-", r3_upper)

r3_losses = r3_values[r3_values < 0].count()
r3_loss_prob = r3_losses / 1000
print("Probabillity of a loss: {:.1%}".format(r3_loss_prob))

Average profit = 6155597.228409678
95% confidence interval : -122184.95237293163 - 12306444.739413176
Probabillity of a loss: 3.0%


I used bootstrapping to calculate the average profit and risk of loss for each region. I found that the risk of a loss for region 3 is to high to justify development. Region one has a lower chance of a loss however the distribution shows that 2.5 % chance that you will only have a profit of only 129,483 USD, and when you consider an investment of 100,000,000 USD that profit margin is not high enough. Region two is the Region with the greatest potential. There is a 97.5% chance that you will have a profit of 1,579,885 USD or higher, and the chance of a loss is only 0.3%. I would reccomend developing the oil wells in region 2.